src/librustc_borrowck/borrowck/mir/dataflow/sanity_check.rs

```diff
@@ -0,0 +1,193 @@
+    // performed by `build_sets`.)
+
+    let blocks = mir.all_basic_blocks();
+    'next_block: for bb in blocks {
+        let bb_data = mir.basic_block_data(bb);
+        let &repr::BasicBlockData { ref statements,
+                                    ref terminator,
+                                    is_cleanup: _ } = bb_data;
+
+        let (args, span) = if let Some(repr::Terminator { ref kind, span, .. }) = *terminator {
+            if let repr::TerminatorKind::Call { func: ref oper, ref args, .. } = *kind
+            {
+                if let repr::Operand::Constant(ref func) = *oper
+                {
+                    if let ty::TyFnDef(def_id, _, &ty::BareFnTy { abi, .. }) = func.ty.sty
+                    {
+                        let name = tcx.item_name(def_id);
+                        if abi == Abi::RustIntrinsic || abi == Abi::PlatformIntrinsic {
+                            if name.as_str() == "rustc_peek" {
+                                (args, span)
+                            } else {
+                                continue;
+                            }
+                        } else {
+                            continue;
+                        }
+                    } else {
+                        continue;
+                    }
+                } else {
+                    continue;
+                }
+            } else {
+                continue;
+            }
+        } else {
+            continue;
+        };
+        assert!(args.len() == 1);
+        let peek_arg_lval = match args[0] {
+            repr::Operand::Consume(ref lval @ repr::Lvalue::Temp(_)) => {
+                lval
+            }
+            repr::Operand::Consume(_) => {
+                bug!("dataflow::sanity_check cannot feed a non-temp to rustc_peek.");
+            }
+            repr::Operand::Constant(_) => {
+                bug!("dataflow::sanity_check cannot feed a constant to rustc_peek.");
+            }
+        };
+
+        let mut entry = results.0.sets.on_entry_set_for(bb.index()).to_owned();
+        let mut gen = results.0.sets.gen_set_for(bb.index()).to_owned();
+        let mut kill = results.0.sets.kill_set_for(bb.index()).to_owned();
+
+        let move_data = flow_ctxt.move_data();
+
+        // Emulate effect of all statements in the block up to (but
+        // not including) the assignment to `peek_arg_lval`. Do *not*
+        // include terminator (since we are peeking the state of the
+        // argument at time immediate preceding Call to `rustc_peek`).
+
+        let mut sets = super::BlockSets { on_entry: &mut entry[..],
+                                          gen_set: &mut gen[..],
+                                          kill_set: &mut kill[..] };
+        // Unfortunately if we just re-do the same thing that dataflow does, then
+        // it will always appear like Lvalues are initialized; e.g. in
+        // a case like:
+        //
+        // <bitset maps var1 to 0>
```

Comment:
I guess the reason *not* to do `rustc_peek(&x)` is that this code runs before borrowck anyway, so e.g. moves don't matter?